main.cpp

```diff
@@ -125,31 +200,145 @@ void uthread_yield(){
     if (ret_val == 1) return;
 
     // Choose the next thread to execute.
-    int id = current_thread_id;
-    do {
-        id = (id + 1) % num_threads;
-        if(id == current_thread_id){
-            printf("infinite loop!\n");
-        }
-    } while (threads[id].complete);
+    ready_list.push_back(current_thread_id);
+
+    current_thread_id = ready_list.front();
+    ready_list.erase(ready_list.begin());
+
+    siglongjmp(threads[current_thread_id].env,1);
+}
+
+/**
+ * Add calling thread to the waiting list to block. Switch to a
+ * different thread.
+ */
+void thread_switch(){
+
+    int ret_val = sigsetjmp(threads[current_thread_id].env,1);
+    if (ret_val == 1) return;
+
+    // Place calling thread on waiting list
+    waiting_list.push_back(current_thread_id);
+
+    // Deadlock
+    if(ready_list.empty()) {
+	printf("Deadlock\n");
+	exit(0);
+    }
+
+    // Take the top thread off the ready list
+    current_thread_id = ready_list.front();
+    ready_list.erase(ready_list.begin());
 
-    // Execute that thread.
-    current_thread_id = id;
     siglongjmp(threads[current_thread_id].env,1);
 }
 
 int uthread_self(){
     return current_thread_id;
 }
 
+int uthread_join(int tid, void **retval){
+    threads[current_thread_id].waiting_for_tid = tid;
+    thread_switch();
+    *retval = threads[tid].result;
+    return 0;
+}
+
+/**
+ * Resume a suspended thread.
+ * @param tid - The tid needed to be resumed
+ */
+int uthread_resume(int tid) {
+    if(tid >= num_threads)
+	return -1;
+    
+    std::vector<int>::iterator it;
+
+    // Verify that tid is currently suspended
+    if((it = find(suspended_list.begin(), suspended_list.end(), tid)) != suspended_list.end()) {
+	suspended_list.erase(it);
+	if(threads[tid].waiting_for_tid == -1)
+	    ready_list.push_back(tid);
+	else
+	    waiting_list.push_back(tid);
+	return 0;
+    }
+
+    // If it wasn't suspended return an error
+    printf("thread_resume: Thread %d was not suspended\n", tid);
+    return -1;
+}
+
+/**
+ * Suspend a thread by adding it to the suspended list
+ * @param - The tid of the thread needed to be suspended
+ */
+int uthread_suspend(int tid) {
+    if(tid >= num_threads)
```

Comment:
I think we should also check that tid is greater than zero, otherwise we could get a segmentation fault.
We should add some unit tests that document this behavior.